src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs

```diff
@@ -52,6 +52,13 @@ public ReferralChasingOptions ReferralChasing
             }
         }
 
+        public void StartTransportLayerSecurity(DirectoryControlCollection controls)
+        {
+            if (_connection._disposed) throw new ObjectDisposedException(GetType().Name);
+            _connection._startTls = true;
+            _connection._startTlsControls = controls;
```

Comment:
Deferring this work also causes a problem with the exception model.

If a server doesn't support StartTLS, or the protocol is still set to 2, or OpenLDAP decides the connection isn't acceptable, or whatever, then the exception needs to come out of this method call, not from the later call to bind, since that's where it would come with Windows.

Reply:
I see. We might be able to send the StartTLS operation manually through OpenLDAP and then call `ldap_install_tls` after the correct URI has been set. But unless that  turns something up, I'm not sure I see a way to preserve the exact same behavior between Windows and Linux using this deferral method. I'll do some more research.

Reply:
instead of deferring the call to startTLS, why not just make sure we call connect() (which will set the URI that we have built so far) before we try to startTLS? I know that with some changes in the options the way that this URI looks might still change, but what is the harm in connecting twice if necessary? By this I mean that if we have the following code:
```c#
LdapConnection connection = new LdapConnection(host);
connection.SessionOptions.ProtocolVersion = 3;
connection.SessionOptions.StartTransportLayerSecurity(null);
connection.SessionOptions.SecureSocketLayer = true;
connection.Bind();
```
We call Connect() to set the URIs twice, once before we start ssl in line 3, and once more in line 5 once we adjust the uri to be ldaps:// instead of ldap://. From what I've tested locally this is ok to do, and the connection and binding still works and is encrypted.

Reply:
I think that's a good idea. Initially, I had avoided that because it is an error to set up TLS twice on the same LDAP handle, but I didn't learn until later that if the URI is changed, then the TLS options are also reset. I will give that a shot.

Reply:
@joperezr, do you know if it's OK to call `Connect()` multiple times on Windows? In the commit I just pushed, I took the conservative route and copied `StartTransportLayerSecurity()` into Windows- and Linux-specific functions, then added these lines to the Linux one:

```csharp
            // Connect first to make sure LDAP URI is set up properly.
            if (!_connection._connected)
            {
                _connection.Connect();
                _connection._connected = true;
            }
```

If it's OK to `Connect()` twice in Windows, I can remove that duplication. (I'm sorry I cannot test it myself.)